Nice tree decomposition for chordal graph

PhoenixIra opened the pull request:
<!-- describe the changes you have made here: what, why, …
     Link issues by using the following pattern: [#333](https://github.com/JabRef/jabref/issues/333).
     The title of the PR must not reference an issue, because GitHub does not support autolinking there. -->
Regarding the issue [#586](https://github.com/jgrapht/jgrapht/issues/586) we created also a nice tree decomposition builder for chordal graphs, which computes a nice tree decomposition with size and time complexity in O(|V|(|V|+|E|)).
We used the idea of an algorithm which outputs a non-nice tree decomposition of size and time complexity in O(|V|+|E|), but adapting it to nice tree decomposition gives here an additional cost (which we do not get for interval graphs).

We also squashed the commit history for a cleaner history and with better readable commits.


----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,266 @@
+ * <li>for every edge $e \in E(G)$, there is a node $t \in V(T)$ with $e \subseteq b(v)$</li>
+ * <li>for all vertices $v \in V(G)$ the set $\{t \in V(T) | v \in b(t)\}$ is non-empty and
+ * connected in $T$</li>
+ * </ul>
```
Let's rephrase this definition a little bit:
```
 * A tree-decomposition of a graph $G = (V, E)$ is a pair $(X, T) = (\{X_i\ |\ i\in I\}, (I, F))$ where
 * $X = \{X_i\ |\ i \in I \}$ is a family of subsets of $V$, and $T = (I, F)$ is a tree, such that
 * <ul>
 * <li>Union of the sets $X_i$ equals to $V$</li>
 * <li>for every edge $e = (u,v)$ there exists a set $X_i$ such that $u \in X_i$ and $v \in X_i$</li>
 * <li>if both $X_i$ and $X_j$ contain a vertex $v$ then every set $X_k$ on the simple path from $X_i$ to $X_j$
 * contains vertex $v$.</li>
 * </ul>
```
This is a verbose variations of the definition introduced in the "Better Algorithms for the Pathwidth and Treewidth of Graphs".

Toptachamann commented on jgrapht-core/src/test/java/org/jgrapht/alg/decompostion/NiceDecompositionBuilderTestUtil.java:
```diff
@@ -0,0 +1,115 @@
+                && union.size() == map.get(second).size()) 
+                    continue; //join node!
+            }
+            assertFalse("Vertex Set "+current+" is not a valid node for a nice decomposition"
```
Can be replaced with ```fail()```

Toptachamann commented on jgrapht-core/src/test/java/org/jgrapht/alg/decompostion/NiceDecompositionBuilderTestUtil.java:
```diff
@@ -0,0 +1,115 @@
+                + "\nin decomposition "+decomposition
+                + "\nwith map "+map, true); //no valid node!
+        }
+        assertTrue(true);
```
Why do we need this?

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,225 @@
+     * @param perfectEliminationOrder the perfect elimination order of the graph
+     * @return a nice decomposition builder for the graph if the graph was chordal, else null
+     */
+    public static <V, E> ChordalNiceDecompositionBuilder<V, E> create(
```
Same here

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,225 @@
+    NiceDecompositionBuilder<V>
+{
+    // the chordal graph
+    Graph<V, E> graph;
```
Is there a reason why the fields aren't private?

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,266 @@
+        decomposition.addEdge(node, vertexChildLeft);
+        decomposition.addEdge(node, vertexChildRight);
+
+        return new Pair<Integer, Integer>(vertexChildLeft, vertexChildRight);
```
You do not need the type parameters. ``return new Pair<>(vertexChildLeft, vertexChildRight);`` also works.

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,266 @@
+     * 
+     * @return unused integer
+     */
+    private Integer getNextInteger()
```
Same as @Toptachamann. Please only use `Integer` when simply `int` won't do.

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,266 @@
+     * @return the newly created node; or null and no change if either {@code forgottenElement} is
+     *         in b({@code node}) or {@code node} is not a leaf.
+     */
+    protected Integer addForget(V forgottenElement, Integer node)
```
As far as I can tell you use the method only once in `ChordalNiceDecompositionBuilder`. There you have this line: `forgetNodeMap.put(vertex, decompNode);` If you return `null` then `vertex` will be removed from the map. Is that the intended behaviour? If this was by design than I guess you can use `Integer` here.

AlexandruValeanu replied:
A better name would be `addForgetNode`.

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,266 @@
+     * 
+     * @param introducedElement the element, which is introduced
+     * @param node the node, which becomes an introduce node
+     * @return the newly created node; or null and no change if either {@code introducedElement} is
```
Use {@code null} for null in the docs.

AlexandruValeanu commented on jgrapht-core/src/test/java/org/jgrapht/alg/decompostion/ChordalityNiceDecompositionBuilderTest.java:
```diff
@@ -0,0 +1,130 @@
+import org.jgrapht.util.SupplierUtil;
+import org.junit.*;
+
+public class ChordalityNiceDecompositionBuilderTest
```
It's a good idea to {@link .....} to the class you are testing.

AlexandruValeanu commented on jgrapht-core/src/test/java/org/jgrapht/alg/decompostion/NiceDecompositionBuilderTestUtil.java:
```diff
@@ -0,0 +1,115 @@
+    private NiceDecompositionBuilderTestUtil() {}
+    
+    /**
+     * tests whether the tree decomposition (decomposition, map) with root is nice
```
Let's rephrase this: **T**ests ... whether the tree decomposition, consisting of the decomposition and the map with the given root is nice. Or something similar.

AlexandruValeanu commented on jgrapht-core/src/test/java/org/jgrapht/alg/decompostion/NiceDecompositionBuilderTestUtil.java:
```diff
@@ -0,0 +1,115 @@
+     */
+    public static <V,E,W> void testNiceDecomposition(Graph<V,E> decomposition, Map<V,Set<W>> map, V root){
+        
+        Queue<V> queue = new LinkedList<V>();
```
Why use a `LinkedList` instead of an `ArrayDeque`?

AlexandruValeanu commented on jgrapht-core/src/test/java/org/jgrapht/alg/decompostion/NiceDecompositionBuilderTestUtil.java:
```diff
@@ -0,0 +1,115 @@
+    }
+    
+    /**
+     * Test whether (decomposition, map) a tree decomposition of oldGraph is
```
Is what? The suspense is killing me. :)

AlexandruValeanu commented on jgrapht-core/src/test/java/org/jgrapht/alg/decompostion/NiceDecompositionBuilderTestUtil.java:
```diff
@@ -0,0 +1,115 @@
+     * @param decomposition the tree decomposition
+     * @param map the map from vertices to the tree decomposition to the sets of the tree decomposition
+     */
+    public static <V,E,W,F> void testDecomposition(Graph<W,F> oldGraph, Graph<V,E> decomposition, Map<V,Set<W>> map){
```
As a general idea, you can make all these functions return a boolean and use `Assert.assertTrue/False` in the actual tests. This way you can test both cases. Your way can only be used to test whether A is B. It doesn't work to test if A is not B.

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,225 @@
+     * @param <V> the vertex type of graph
+     * @param <E> the edge type of graph
+     * @param graph the chordal graph for which a decomposition should be created
+     * @return a nice decomposition builder for the graph if the graph was chordal, else null
```
If you are going with @Toptachamann, which I believe is a good idea for this class, don't forget to add `@throws` in the javadoc to explain why it may throw an `IllegalArgumentException`.

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,225 @@
+            return null;
+        ChordalNiceDecompositionBuilder<V, E> builder =
+            new ChordalNiceDecompositionBuilder<>(graph, inspec.getPerfectEliminationOrder());
+        builder.computeNiceDecomposition();
```
You should mention that it doesn't simply return a builder but it also builds the nice decomposition. The user might expect it to be lazy which is not the case here. In the case of builders, it is a good idea to mention whether the builder actually does something when created.

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,225 @@
+     * @param <E> the edge type of graph
+     * @param graph the chordal graph for which a decomposition should be created
+     * @param perfectEliminationOrder the perfect elimination order of the graph
+     * @return a nice decomposition builder for the graph if the graph was chordal, else null
```
{@code null}

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,225 @@
+
+    /**
+     * Factory method for the nice decomposition builder of chordal graphs. This method needs the
+     * perfect elimination order. It does not check whether the order is correct. This method may
```
Extract this information in a separate line. "Note: This method does NOT check whether the order is correct". And have this is a separate paragraph or at least a separate line.  "This method may behave **arbitrarily** if the perfect elimination order is incorrect" is not actually needed as it doesn't provide any extra information for the caller.

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,225 @@
+     */
+    private void computeNiceDecomposition()
+    {
+
```
Extra line.

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,225 @@
+
+        // map from vertices to decomposition-nodes where the decomposition-node has all its
+        // predecessors
+        Map<V, Integer> forgetNodeMap = new HashMap<V, Integer>(graph.vertexSet().size());
```
You can make use of type inference and have `HashMap<>(...)`

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,225 @@
+            Set<V> clique = new HashSet<V>(predecessors);
+            clique.add(vertex);
+            Set<V> toIntroduce = new HashSet<V>(decompositionMap.get(decompNode));
+            toIntroduce.removeAll(clique);
```
Or simply have `toIntroduce.removeAll(predecessors); toIntroduce.remove(vertex)`. Why do you need to create a new set?

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,225 @@
+        }
+        //finish all unfinished paths
+        leafClosure();
+
```
Extra line.

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,225 @@
+    private Set<V> getOrderPredecessors(Map<V, Integer> map, V vertex)
+    {
+        Set<V> predecessors = new HashSet<>();
+        Integer vertexPosition = map.get(vertex);
```
int

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,225 @@
+        Set<E> edges = graph.edgesOf(vertex);
+        for (E edge : edges) {
+            V oppositeVertex = Graphs.getOppositeVertex(graph, edge, vertex);
+            Integer destPosition = map.get(oppositeVertex);
```
int

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,266 @@
+ * </ul>
+ * </ul>
+ * <p>
+ * See:<br>
```
See ... for a more complete ...

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,266 @@
+    }
+
+    /**
+     * Getter for the next free Integer. Supplies the add vertex methods with new vertices
```
Define free

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,266 @@
+    /**
+     * Getter for the next free Integer. Supplies the add vertex methods with new vertices
+     * 
+     * @return unused integer
```
Use either free or unused or next. And make sure it is clear what you mean by it.

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,266 @@
+     * @return the newly created node; or null and no change if either {@code introducedElement} is
+     *         in b({@code node}) or {@code node} is not a leaf.
+     */
+    protected Integer addIntroduce(V introducedElement, Integer node)
```
A better name would be addIntroduceNode.

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,266 @@
+    }
+
+    /**
+     * Returns the map from integer nodes of the tree decomposition {@code getDecomposition()} to
```
Returns an unmodifiable map from ....

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,266 @@
+     * Returns the map from integer nodes of the tree decomposition {@code getDecomposition()} to
+     * the sets of the vertices from the graph as an unmodifiable map
+     * 
+     * @return the nodes of decomposition to sets of vertices map
```
the map from nodes of decomposition to sets of vertices

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,266 @@
+    /**
+     * Returns the root of the decomposition {@code getDecomposition()}
+     * 
+     * @return the root the decomposition
```
the root **of** the decomposition

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,266 @@
+    }
+
+    /**
+     * Returns the root of the decomposition {@code getDecomposition()}
```
Get the root of the decomposition computed by {@code getDecomposition()}

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,266 @@
+     * 
+     * @return the root the decomposition
+     */
+    public Integer getRoot()
```
int

AlexandruValeanu commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/package-info.java:
```diff
@@ -0,0 +1,6 @@
+
+/**
+ * Tree Decomposition related algorithms
```
Clash with existing package-info. There's no need to change it.

AlexandruValeanu commented on jgrapht-core/src/test/java/org/jgrapht/alg/decompostion/ChordalityNiceDecompositionBuilderTest.java:
```diff
@@ -0,0 +1,130 @@
+package org.jgrapht.alg.decompostion;
```
Where's the header?

AlexandruValeanu commented on jgrapht-core/src/test/java/org/jgrapht/alg/decompostion/NiceDecompositionBuilderTestUtil.java:
```diff
@@ -0,0 +1,115 @@
+package org.jgrapht.alg.decompostion;
```
Where's the header?

PhoenixIra commented:
I used non-open lecture notes, where nice decomposition was defined with leaves and roots with empty sets. But since it was not difficult to change it, i did so.
Furthermore I also changed the factory methods to constructors and merged master into the branch (and squashed my commits)

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,291 @@
+{
+
+    // resulting decomposition
+    protected Graph<Integer, DefaultEdge> decomposition;
```
I am wondering why not use directly something like `Graph<Set<V>, DefaultEdge>` for the decomposition (and for the result of the decomposition).

PhoenixIra replied:
This was also my first attempt, but a decomposition (especially a nice decomposition) can multiple nodes with same sets. Just look at the definition of a join node, where the join node and both of their children have the same set. But the graph class (or more precise: the vertex set class) would consider these three nodes the same. So my two options was to either use a Set with modified equal/hashvalue getter, a wrapper class or a Map from Node to Sets. Since the last one is equal to the formal definition, I opt for this method.

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,210 @@
+        this.graph = graph;
+        this.perfectOrder = inspec.getPerfectEliminationOrder();
+        vertexInOrder = getVertexInOrder();
+        computeNiceDecomposition();
```
Should we run the algorithm at the construction time or after the first call to some getter method? @jkinable @jsichi @d-michail what do you think?

Toptachamann replied:
Since a decomposition of a graph and a tree decomposition of a graph are different things, it’s better to name it ChordalNiceTreeDecomposition and NiceTreeDecompositionBase

jsichi replied:
Good point.

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,210 @@
+    private void computeNiceDecomposition()
+    {
+        // map from vertices to decomposition-nodes where the decomposition-node has all its
+        // predecessors
```
Let's mention that here the decomposition nodes are forget nodes and they also contain corresponding vertices:
```
/*
 * Map from the vertices of the graph to the forget nodes from the decomposition tree. The mapping 
 * is one-to-one and each decomposition node contains corresponding vertex and all its predecessors.
 */
```

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,210 @@
+        // predecessors
+        Map<V, Integer> forgetNodeMap = new HashMap<>(graph.vertexSet().size());
+
+        //iterate over the perfect order
```
`// iterator over the perfect elimination order`

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,210 @@
+            // get the predecessors regarding the perfect elimination order
+            List<V> predecessors = getOrderPredecessors(vertexInOrder, vertex);
+
+            // calculate nearest successors according to perfect elimination order
```
nearest predecessor

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,210 @@
+                    lastVertex = predecessor;
+            }
+
+            // get node with clique of last vertex, else we use the last node
```
If the current vertex has no predecessors, then we can use any decomposition node to build a path to, can't we?

PhoenixIra replied:
Correct. Thats why I implicial just use the last one here.

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,210 @@
+                decompNode = addJoinNode(decompNode).getFirst();
+            }
+
+            // calculate vertices of nearest successor, which needs to be handled
```
What is the nearest successor?

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,291 @@
+* A tree-decomposition of a graph $G = (V, E)$ is a pair $(X, T) = (\{X_i\ |\ i\in I\}, (I, F))$ where
+ * $X = \{X_i\ |\ i \in I \}$ is a family of subsets of $V$, and $T = (I, F)$ is a tree, such that
+ * <ul>
+ * <li>Union of the sets $X_i$ equals to $V$</li>
```
Let's make it lower case: Union -> union

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,291 @@
+        // create first/left new child of node
+        int vertexChildLeft = getNextInteger();
+        decomposition.addVertex(vertexChildLeft);
+        currentVertexBag = new HashSet<V>(decompositionMap.get(node));
```
Please, remove generic type parameters on the right-hand side of the assignments.

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,291 @@
+        // create second/right new child of node
+        int vertexChildRight = getNextInteger();
+        decomposition.addVertex(vertexChildRight);
+        currentVertexBag = new HashSet<V>(decompositionMap.get(node));
```
Same here

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,291 @@
+     */
+    protected void leafClosure()
+    {
+        Set<Integer> vertices = new HashSet<Integer>(decomposition.vertexSet());
```
and here

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,291 @@
+ * <li>if both $X_i$ and $X_j$ contain a vertex $v$ then every set $X_k$ on the simple path from $X_i$ to $X_j$
+ * contains vertex $v$.</li>
+ * </ul>
+ * <br>
```
Let's mention the mapping `b` or replace its usages below with the sets X_i. I would choose the first alternative since it is convenient to think of the decomposition tree as the mapping from tree nodes to the sets containing vertices from the initial graph.

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,291 @@
+ * An abstract builder class for nice tree decompositions, which builds the tree decomposition in a
+ * top-down manner.
+ * <p>
+* A tree-decomposition of a graph $G = (V, E)$ is a pair $(X, T) = (\{X_i\ |\ i\in I\}, (I, F))$ where
```
Add space at the beginning

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,291 @@
+import org.jgrapht.graph.*;
+
+/**
+ * An abstract builder class for nice tree decompositions, which builds the tree decomposition in a
```
An abstract class for building a nice tree decomposition in a top-down manner.

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java:
```diff
@@ -0,0 +1,210 @@
+ * <p>
+ * The complexity of this algorithm is in $\mathcal{O}(|V|(|V|+|E|))$.<br>
+ * Consider the every node in the nice tree decomposition: There are exactly $|V|$ many forget
+ * nodes. There are at most $2|V|$ additionally nodes because of a join nodes. Every join node
```
Probably, exactly |V| - 1 forget nodes since for the root we do not add a forget node. And are there at most 2|V| - 2 auxiliary nodes produced by adding join nodes?

PhoenixIra replied:
Correct, yes. First one is due to changing the condition from |b(root)|=0 to |b(root)|=1, second one was just an overapproximation, but I changed both accordingly.

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceTreeDecomposition.java:
```diff
@@ -0,0 +1,215 @@
+    private Map<V, Integer> vertexInOrder;
+
+    /**
+     * Constructor for the nice decomposition builder of chordal graphs. The constructor
```
Constructor for the nice tree decomposition algorithm

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceTreeDecomposition.java:
```diff
@@ -0,0 +1,215 @@
+    }
+
+    /**
+     * Constructor for the nice decomposition builder of chordal graphs. 
```
Same here

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceTreeDecomposition.java:
```diff
@@ -0,0 +1,215 @@
+    }
+
+    /**
+     * Computes the nice decomposition of the graph. We iterate over the perfect elimination order
```
nice decomposition -> nice tree decomposition

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceTreeDecomposition.java:
```diff
@@ -0,0 +1,215 @@
+ * introduced vertex. Since this introduced vertex is part of the clique of the least ancestor
+ * forget node, the corresponding bag of the introduce node is smaller than the set of neighbors of
+ * this vertex. Thus the time complexity is in $\mathcal{O}(|V|(|V|+|E|))$.
+ * <p>
```
We generate `|V| - 1` forget nodes and build `|V|-1` paths to previously added nodes in the tree. There is at most one join node generated per such path, therefore there are at most `|V|-1` join nodes. At least we are calling addJoinNode once per vertex from the perfect elimination order. 

I don't quite understand why the running time of the algorithm is O(V(V + E)). Computing order predecessors and finding one with the highest index contributes O(V + E) overall. Creating and adding forget nodes run in O(V+ E) since the sum of the cardinalities of all predecessors sets equals to |E|. I'm uncertain about the running time complexity of deleting the nodes while building and finishing the paths. Every vertex can be O(deg(v)) times the higher predecessor, consequently it can be deleted O(deg(v)) times (deleting a node = adding "introduce" node). Deleting a node takes O(deg(v)) time since we a dealing with bags which contain cliques. This is done for every vertex, that's why the running time is (V*vertex_max_deg^2 + E) which is O(VE^2) worst case.  Can we have a better complexity analysis? 

Another point is that we have O(V^2) upper bound on introduce nodes, but adding one introduce node runs in O(vertex_max_deg) time because of set copying. So, the running time is at least O(V^2*vertex_max_deg). Can we aggregate this somehow?

PhoenixIra replied:
I do not quite get your first point. But for the second one, I do not agree. Adding one introduce node runs in O(|b(parent)|) - which is important, since in this case, the parent consits of a clique. Thus adding an introduce node for the vertex v costs in this case O(|Neighbour(v)|). Which is far less then O(vertex_max_deg). But I guess, I can try to do another analysis. Maybe this one is more elegant.

First lets proof the following lemma: every bag of a node in the tree decomposition consists of a clique. Proof by induction: the roots bag is obviously a (trivial) clique. Additionally, adding a forget node produces a node with a clique by construction, because the bag consists of the forgotten element and its predecessors. 
Adding join nodes and adding introduce nodes only adds nodes, which bags are a subset of their parents bag. Since every subset of a clique is a clique, their bags are a clique by induction hypothesis. 

Now there are at most |V| different pathes from the root to a leaf, because every vertex can create at most one new path. Consider any of them. For now we ignore the join nodes. Because the tree is a tree decomposition, every vertex v can be introduced and forgotten at most ones. Adding a forget node generates a node with a clique of v. Adding an introduce node, generates a node with a clique of v, whithout v (because v is in the bag of its parent and the parents bag is a clique). Thus the cost of adding a forget node or adding an introduce node for one v is in O(|Neighbour(v)|). Now every vertex has at most one introduce node and one forget node on this path, which adds up to O(|V|+|E|) for one path. Since there are at most |V| different paths, we have O(|V|(|V|+|E|)).

Now considering the join nodes: There are 2*(|V|-1) join nodes, everyone can be generated with at most vertex_max_deg sized bags, thus they add up to O(|V|*vertex_max_deg). Simmilar calculating the toIntroduce sets add up to O(|V|*vertex_max_deg). In total again O(|V|(|V|+|E|)).

Toptachamann replied:
Now the analysis is clear for me, thanks! One more issue to resolve: why there are 2*(V-1) joins (at most)? We are adding one join node per forget node, consequently there are at most (V-1) join nodes.

PhoenixIra replied:
You are right. There are (V-1) join nodes, but adding join nodes generate 2 nodes. But these two nodes are not (necessary) join nodes according to the definition.

Toptachamann replied:
All right, thanks! Let's use the following Javadoc:
```
/**
 * Builds a nice tree decomposition of chordal graphs. See {@link NiceTreeDecompositionBase} for
 * an explanation of the nice tree decomposition.
 * <p>
 * For a given perfect elimination order of the graph $G = (V, E)$ every vertex in $V(G)$ has an associated 
 * index in this order. The predecessors of a vertex $x \in V(G)$ are those vertices incident to $x$ that 
 * have smaller index than $x$ (a vertex can have no predecessors). This class uses a perfect elimination 
 * order produced by {@link ChordalityInspector} to iteratively add vertices of $G$ to some nodes in the 
 * tree $T$. Initially, it adds a bag to the decomposition with the first vertex from the perfect elimination 
 * order as a root of the tree. Then for all remaining vertices $v \in V(G)$ this class generates a node (bag) 
 * $s \in I(T)$ in the tree $T$ which contains vertex $v$ and all its predecessors. Then it builds a path from 
 * $s \in I(T)$ to $t \in I(T)$, where $s$ is the node where the predecessor of $v$ with the highest index was 
 * added to the decomposition for the first time. If the vertex $v$ has no predecessors, a path is built to an 
 * arbitrary node from the tree $T$. The bags on the resulting path between $s$ and $t$ contain only cliques.
 * <p>
 * The time complexity of this algorithm is $\mathcal{O}(|V|(|V|+|E|))$. In the decomposition produced by this 
 * class there are exactly $|V|-1$ many forget nodes since they are generated for every vertex in the $V(G)$ 
 * except for the vertex added as a root. There are at most $|V|-1$ additionally added join nodes which are 
 * generated during the connection of forget nodes to the tree. Every join node creates one additional 
 * root-to-leaf path. Therefore, there are $\mathcal{O}(|V|)$ root-to-leaf paths in the resulting decomposition. 
 * Every root-to-left path can contain $\mathcal{O}(|V|)$ introduce nodes, which yields $\mathcal{O}(|V|^2)$ 
 * upper bound on introduce nodes. Now considering the bags of the introduce nodes. Bags of every root-to-leaf 
 * path can contain at most $|V|$ distinct vertices of the initial graph. Every vertex $v$ can be introduced and 
 * forgotten on a path at most once. The sizes of these introduce and forget nodes are $\mathcal{O}(deg(v) + 1)$. 
 * Consequently, every root-to-leaf path is constructed in $\mathcal{O}(|V| + |E|)$ time. Since there are 
 * $\mathcal{O}(|V|)$ root-to-leaf paths, the total running time is $\mathcal{O}(|V|(|V| + |E|))$.
 * <p>
 * The algorithm used to construct a nice tree decomposition is a non-recursive adaption of algorithm 2 
 * from here: <br>
 * Hans L. Bodlaender, Arie M.C.A. Koster, Treewidth computations I. Upper bounds, Information and
 * Computation, Volume 208, Issue 3, 2010, Pages 259-275,
 * 
 * @author Ira Justus Fesefeldt (PhoenixIra)
 * @author Timofey Chudakov
 *
 * @since June 2018
 *
 * @param <V> the vertex type of the graph
 * @param <E> the edge type of the graph
 */
```
I've used your time analysis and mentioned a few relevant definitions.

PhoenixIra replied:
Thanks for the javadoc! I changed a small part regarding the forget nodes, which I came up with after you mentioned the join nodes: join nodes double existing nodes and adds a join node. Thus we can have 2(|V|-1) many forget nodes by doubling every forget node ones:

/**
 * Builds a nice tree decomposition of chordal graphs. See {@link NiceTreeDecompositionBase} for an
 * explanation of the nice tree decomposition.
 * <p>
 * For a given perfect elimination order of the graph $G = (V, E)$ every vertex in $V(G)$ has an
 * associated index in this order. The predecessors of a vertex $x \in V(G)$ are those vertices
 * incident to $x$ that have smaller index than $x$ (a vertex can have no predecessors). This class
 * uses a perfect elimination order produced by {@link ChordalityInspector} to iteratively add
 * vertices of $G$ to some nodes in the tree $T$. Initially, it adds a bag to the decomposition with
 * the first vertex from the perfect elimination order as a root of the tree. Then for all remaining
 * vertices $v \in V(G)$ this class generates a node (bag) $s \in I(T)$ in the tree $T$ which
 * contains vertex $v$ and all its predecessors. Then it builds a path from $s \in I(T)$ to $t \in
 * I(T)$, where $s$ is the node where the predecessor of $v$ with the highest index was added to the
 * decomposition for the first time. If the vertex $v$ has no predecessors, a path is built to an
 * arbitrary node from the tree $T$. The bags on the resulting path between $s$ and $t$ contain only
 * cliques.
 * <p>
 * The time complexity of this algorithm is $\mathcal{O}(|V|(|V|+|E|))$. In the decomposition
 * produced by this class there are at most $|V|-1$ additionally added join nodes which are
 * generated during the connection of forget nodes to the tree. The algorithm adds exactly $|V|-1$
 * many forget nodes since they are generated for every vertex in the $V(G)$ except for the vertex
 * added as a root. They may be doubled by join nodes, which adds up to at most $2(|V|-1)$ forget
 * nodes. Every join node creates one additional root-to-leaf path. Therefore, there are
 * $\mathcal{O}(|V|)$ root-to-leaf paths in the resulting decomposition. Every root-to-left path can
 * contain $\mathcal{O}(|V|)$ introduce nodes, which yields $\mathcal{O}(|V|^2)$ upper bound on
 * introduce nodes. Now considering the bags of the introduce nodes. Bags of every root-to-leaf path
 * can contain at most $|V|$ distinct vertices of the initial graph. Every vertex $v$ can be
 * introduced and forgotten on a path at most once. The sizes of these introduce and forget nodes
 * are $\mathcal{O}(deg(v) + 1)$. Consequently, every root-to-leaf path is constructed in
 * $\mathcal{O}(|V| + |E|)$ time. Since there are $\mathcal{O}(|V|)$ root-to-leaf paths, the total
 * running time is $\mathcal{O}(|V|(|V| + |E|))$.
 * <p>
 * The algorithm used to construct a nice tree decomposition is a non-recursive adaption of
 * algorithm 2 from here: <br>
 * Hans L. Bodlaender, Arie M.C.A. Koster, Treewidth computations I. Upper bounds, Information and
 * Computation, Volume 208, Issue 3, 2010, Pages 259-275,
 * 
 * @author Ira Justus Fesefeldt (PhoenixIra)
 * @author Timofey Chudakov
 *
 * @since June 2018
 *
 * @param <V> the vertex type of the graph
 * @param <E> the edge type of the graph
 */

Toptachamann replied:
That’s right

Toptachamann commented on jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/ChordalityNiceTreeDecompositionTest.java:
```diff
@@ -0,0 +1,167 @@
+ *
+ * @author Ira Justus Fesefeldt (PhoenixIra)
+ */
+public class ChordalityNiceTreeDecompositionTest
```
Since the class that is being tested is `ChordalNiceTreeDecomposition`, this one should be named `ChordalNiceTreeDecompositionTest`

Toptachamann commented on jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/NiceTreeDecompositionTestUtil.java:
```diff
@@ -0,0 +1,143 @@
+        Queue<V> queue = new ArrayDeque<>();
+        
+        //test and add root
+        if(map.get(root).size() != 1)
```
I think we should use braces with the `if` statements when their body consists of a single line. So, please, add braces in the single line `if` statements.

PhoenixIra commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceTreeDecompositionBase.java:
```diff
@@ -0,0 +1,319 @@
+ *
+ * @param <V> the vertices of the graph
+ */
+abstract public class NiceTreeDecompositionBase<V>
```
Most papers reference to an algorithm which can transform every graph G with treewidth k into a nice tree decomposition of width k. (Kloks, Ton: Treewidth - Computations and Approximations, Lecture Notes in Computer Science, 1994, p.149, Lemma 13.1.2). The case distinction looks similar to the methods of this base class to me.

Additionally I have already implemented an algorithm for interval graphs (as an interval representation) using the base class. The idea of the algorithm is briefly mentioned in this paper (Fomin F.V., Heggernes P., Mihai R., Mixed search number and linear-width of interval and split graphs, In: Brandstädt A., Kratsch D., Müller H. (eds), Graph-Theoretic Concepts in Computer Science, 2007, p.306).

PhoenixIra commented:
I think a class TreeDecomposition<V, I, E> might be useful. Giving the option to change the node type of the tree decomposition can also be useful: especially for nice tree decompositions you may want to have nodes of the type forget, introduce and join instead of checking it manually with set operations. 

But where should this class be located?
